gcc/cmassoc/matrix.1
.TH matrix 1 "August 2013" "cmassoc-tools-1.9.1" "Motley Tools"

.SH NAME
matrix - print parse matrix table

.SH SYNOPSIS
.BR matrix
.RI [ options ]
.IR file 
.RI [ file ]
.RI [ ... ] 
.RI [> stdout ]

.PP
.BR matrix
.RI [ options ]
.RI [< stdin ]
.RI [> stdout ]

.SH DESCRIPTION
Read a matrix definition file and print an HTML or DOCBOOK table on stdout.
The output is suitable for use in the body of another HTML page or DOCBOOK section.
Use program \fBhtml\fR to embed HTML output from this program inside a template HTML page.

.PP
A matrix description file is similar to a comma-separated-value file but spaces are used to separate field and a semi-colon is used to separate tuples.
The first line of the file defines table column headings.

.SH OPTIONS
Long options are not supported.
All options are single letters.

.TP
.RB - d
Print a table on stdout using DOCBOOK elements \fBtable\fR, \fBtitle\fR, \fBrow\fR and \fBentry\fR.
The indent level can be varied using option \fB-i\fR.

.TP
-\fBe\fI string\fR
The string displayed inside empty table cells.
The default is "\fB&nbsp;\fR" which displays as an empty cell.

.TP
.RB - h
Print table on stdout using HTML elements \fBtable\fR, \fBtr\fR and \fBtd\fR.
The indent level can be varied using option \fB-i\fR.

.TP
-\fBl\fI level\fR
The starting indent level.
This option may be used to match the indent level of companion documents or document fragments.
Indents are implemented using horizontal tabs.
Valid values are \fB0\fR through \fB16\fR.
The default is \fB0\fR.

.TP
-\fBe\fI string\fR
The string displayed inside matching cells.
This option is useful for short tables where the table header does not scroll off the screen.
The default is the column title.

.TP
.RB - q
Quiet mode.
This option has no effect at this time.

.TP
.RB - s
Print a CSS stylesheet on stdout where class names match those in the HTML table.
The stylesheet may be used to create new CSS stylesheets, included in existing CSS stylesheets or included in the HTML section.

.TP
.RB - v
Verbose mode.
This option has no effect at this time.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over previous command line options.

.TP
.RB - ! ,-- version
Print program version information on stdout.
This option takes precedence over previous command line options.

.SH ARGUMENTS

.TP
\fIfile\fR
The name of a matrix definition file.
A matrix definition file consists of one table column definition and many table row definitions.
Table column and row definitions have the same format but distinguished by file position.
The first is the table column definition while successive ones are row definitions.
A definition consists of one or more fields separated by white space.
Definitions are separated from each other by one semi-colon.
This means that a definition can straddle multiple lines and that a semi-colon must appear between the last field of a definition and the first field of the next definition.
A semicolon need not appear after the last field of the last definition in a file and should be omitted to avoid creating an empty table row at the bottom of the output.

.SH EXAMPLES
An example matrix definition file, \fBmatrix.txt\fR, appears below.
The first line is the table column definition.
The remaining lines are table row definitions.
There are 4 columns and 3 rows defined here.
Names are separated by white space and each definition, except the last, is terminated by semicolon.
Names may contain white space and puntuation when enclosed in quotes.

.PP
   Items List1 List2 List3;
   Item1 List2; 
   Item2 
     List1
     List3;
   Item3 List3
     List2
   
     List1

.PP
The following command reads file \fBmatrix.txt\fR (shown above) and writes an HTML table to stdout because HTML is the default output format.

.PP
   # matrix < matrix.txt > matrix.html

.PP

.PP
File \fBmatrix.html\fR will look something like this when viewed using a web browser.
There will be additional formatting, of course, depending on browser and stylesheet settings.

.PP 
   Title
   Items List1 List2 List3
   Item1       Yes	  
   Item2 Yes         Yes
   Item3 Yes   Yes   Yes

.PP
File \fBmatrix.html\fR will look like this when opened using a standard text editor.
See \fBhttp://www.w3.org/MarkUp/\fR for more information on HTML.

.PP
   <h1 class='matrix'>
   	Title
   	</h1>
   <table class='matrix'>
   	<tr class='matrix'>
   		<th class='matrix'>
   			Items
   			</th>
   		<th class='matrix'>
   			List1
   			</th>
   		<th class='matrix'>
   			List2
   			</th>
   		<th class='matrix'>
   			List3
   			</th>
   		</tr>
   	<tr class='matrix'>
   		<td class='matrix'>
   			Item1
   			</td>
   		<td class='empty'>
   			&nbsp;
   			</td>
   		<td class='match'>
   			Yes
   			</td>
   		<td class='empty'>
   			&nbsp;
   			</td>
   		</tr>
   	<tr class='matrix'>
   		<td class='matrix'>
   			Item2
   			</td>
   		<td class='match'>
   			Yes
   			</td>
   		<td class='empty'>
   			&nbsp;
   			</td>
   		<td class='match'>
   			Yes
   			</td>
   		</tr>
   	<tr class='matrix'>
   		<td class='matrix'>
   			Item3
   			</td>
   		<td class='match'>
   			Yes
   			</td>
   		<td class='match'>
   			Yes
   			</td>
   		<td class='match'>
   			Yes
   			</td>
   		</tr>
   	</table>

.PP
The next example pipes program output into program \fBhtml\fR to bracket the output with an HTML header and footer.
Option \fB-h\fR is present to ensure output is in HTML format.
Option \fB-l\fR is used to adjust the indent level to match that expected of program \fBhtml\fR.
The result is saved in file \fBmatrix.html\fR.

.PP
   # matrix -h -l2 matrix.txt | html > matrix.html

.PP
The next example reads file \fBmatrix.txt\fR and writes a DOCBOOK table to stdout because option \fB-d\fR is present.

.PP
   # matrix -d < matrix.txt > matrix.xml

.PP
File \fBmatrix.xml\fR will look like this when opened using a standard text editor.
The element names are different thatn HTML but the basic principle is the same.
See \fBhttp://www.docbook.org\fR for more information on DOCBOOK.

.PP
   <table>
   	<title>
   		Title
   		</title>
   	<tgroup>
   		<thead>
   			<row>
   				<entry>
   					Items
   					</entry>
   				<entry>
   					List1
   					</entry>
   				<entry>
   					List2
   					</entry>
   				<entry>
   					List3
   					</entry>
   				</row>
   			<thead>
   		<tbody>
   			<row>
   				<entry>
   					Item1
   					</entry>
   				<entry>
   					&nbsp;
   					</entry>
   				<entry>
   					Yes
   					</entry>
   				<entry>
   					&nbsp;
   					</entry>
   				</row>
   			<row>
   				<entry>
   					Item2
   					</entry>
   				<entry>
   					Yes
   					</entry>
   				<entry>
   					&nbsp;
   					</entry>
   				<entry>
   					Yes
   					</entry>
   				</row>
   			<row>
   				<entry>
   					Item3
   					</entry>
   				<entry>
   					Yes
   					</entry>
   				<entry>
   					Yes
   					</entry>
   				<entry>
   					Yes
   					</entry>
   				</row>
   			</tbody>
   		</tgroup>
   	</table>

.SH SEE ALSO
.BR html ( 1 ),
.BR offset ( 1 )

.SH CREDITS
 Charles Maier <cmaier@cmassoc.net>



gcc/cmassoc/matrix.c
/*====================================================================r
 *
 *   matrix.c - 
 *
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

#define _GETOPT_H

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/version.h"
#include "../tools/number.h"
#include "../tools/markup.h"
#include "../tools/format.h"
#include "../tools/error.h"
#include "../tools/types.h"
#include "../tools/sizes.h"
#include "../tools/chars.h"
#include "../linux/linux.h"
#include "../date/date.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/indent.c"
#include "../tools/error.c"
#include "../tools/efreopen.c"
#include "../tools/uintspec.c"
#include "../tools/todigit.c"
#endif

#ifndef MAKEFILE
#include "../linux/hostname.c"
#include "../linux/username.c"
#include "../linux/getusername.c"
#endif

/*====================================================================*
 *   program constants;
 *--------------------------------------------------------------------*/

#define LEVEL 0
#define EMPTY "&nbsp;"
#define MATCH ""
#define TITLE "Matrix"

/*====================================================================*
 *   program variables;
 *--------------------------------------------------------------------*/

typedef struct column 

{ 
	struct column * next; 
	char const * style; 
	char const * label; 
	char * name; 
} 

column; 
static unsigned line = 1; 
static unsigned cols = 0; 
static char string [_NAMESIZE] = ""; 
static char const * style = "matrix"; 
static char const * style_posted = "posted"; 
static char const * style_empty = "empty"; 
static char const * style_match = "match"; 
static char const * label_empty = EMPTY; 
static char const * label_match = MATCH; 

/*====================================================================*
 *
 *   signed gettoken (char * string);
 *
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static signed gettoken (char * string) 

{ 
	extern unsigned line; 
	signed c; 
	do 
	{ 
		c = getc (stdin); 
		if (c == '\n') 
		{ 
			line++; 
		} 
	} 
	while (isspace (c)); 
	if (isalpha (c) || (c == '_')) 
	{ 
		do 
		{ 
			* string++ = c; 
			c = getc (stdin); 
		} 
		while (isalnum (c) || (c == '.') || (c == ':') || (c == '_') || (c == '-')); 
		ungetc (c, stdin); 
		c = 'A'; 
	} 
	else if (isdigit (c)) 
	{ 
		do 
		{ 
			* string++ = c; 
			c = getc (stdin); 
		} 
		while (isdigit (c) || (c == '.')); 
		ungetc (c, stdin); 
		c = '0'; 
	} 
	else if (isquote (c)) 
	{ 
		char o = getc (stdin); ; 
		while ((o != c) && (c != EOF)) 
		{ 
			if (o == '\\') 
			{ 
				o = getc (stdin); 
			} 
			* string++ = o; 
			o = getc (stdin); 
		} 
	} 
	else 
	{ 
		* string++ = c; 
	} 
	* string++ = (char) (0); 
	return (c); 
} 

/*====================================================================*
 *
 *   static void headers (struct column * column);
 *
 *   read first file statement and store column names;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void headers (struct column * column) 

{ 
	extern char const * style_empty; 
	extern char const * label_empty; 
	extern char string []; 
	if (~ gettoken (string)) 
	{ 
		column->next = (struct column *) (0); 
		column->name = strdup (string); 
		column->label = label_empty; 
		column->style = style_empty; 
		for (cols = 0; ~ gettoken (string) && (* string != ';'); cols++) 
		{ 
			column->next = (struct column *)(malloc (sizeof (struct column))); 
			column = column->next; 
			column->next = (struct column *) (0); 
			column->name = strdup (string); 
			column->label = label_empty; 
			column->style = style_empty; 
		} 
	} 
	return; 
} 

/*====================================================================*
 *
 *   static void columns (struct column * column);
 *
 *   read next file statement to store row names and match column
 *   headers;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void columns (struct column * column) 

{ 
	extern char const * style_match; 
	extern char const * style_empty; 
	extern char const * label_match; 
	extern char const * label_empty; 
	extern char string []; 
	if (~ gettoken (string)) 
	{ 
		strcpy (column->name, string); 
		while (~ gettoken (string) && (* string != ';')) 
		{ 
			struct column * object; 
			for (object = column->next; object; object = object->next) 
			{ 
				if (!strcmp (object->name, string)) 
				{ 
					object->style = style_match; 
					object->label = * label_match? label_match: object->name; 
					break; 
				} 
			} 
			if (!object) 
			{ 
				error (0, 0, "%s \"%s\" on line %d is an orphan", column->name, string, line); 
			} 
		} 
	} 
	return; 
} 

/*====================================================================*
 *
 *   void finish (struct column * column);
 *
 *   free memory used to format table;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static void finish (struct column * column) 

{ 
	column->style = 0; 
	column->label = 0; 
	free (column->name); 
	while (column->next) 
	{ 
		struct column * object = column->next; 
		column->next = object->next; 
		object->next = (struct column *)(0); 
		object->style = 0; 
		object->label = 0; 
		free (object->name); 
		free (object); 
	} 
	return; 
} 

/*====================================================================*
 *
 *   static unsigned stylesheet (unsigned margin);
 *
 *   print CSS2 stylesheet for HTML page;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned stylesheet (unsigned margin) 

{ 
	indent (margin, "table { table-layout: fixed; background: transparent; border: solid 1pt black; border-collapse: separate; border-spacing: 1px; font: normal 10pt verdana; }"); 
	indent (margin, "th { background: inherit; border: solid 1pt black; padding: 2px 10px; text-align: center; vertical-align: top; color: maroon; }"); 
	indent (margin, "td { background: inherit; border: solid 1pt black; padding: 2px 10px; text-align: left; vertical-align: top; color: navy; }"); 
	indent (margin, "td.%s { background: #FFFF00; }", style_match); 
	indent (margin, "td.%s { background: #FFFFFF; }", style_empty); 
	indent (margin, "div.%s { text-align: center; margin: 10px; };", style_posted); 
	return (margin); 
} 

/*====================================================================*
 *
 *   unsigned posted ();
 *
 *   output posting date and time plus username and hostname;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned posted (unsigned margin) 

{ 
	time_t now = time (& now); 
	static char datetime [LOGTIME_LEN]; 
	strftime (datetime, sizeof (datetime), LOGTIME, localtime (& now)); 
	indent (margin++, "<div class='%s'>", style_posted); 
	indent (margin, "Posted %s on %s by %s", datetime, hostname (), username (getuid ())); 
	indent (margin--, "</div>"); 
	return (margin); 
} 

/*====================================================================*
 *
 *   unsigned table1 (unsigned margin, char const * string, struct column * column);
 *
 *   print table in HTML format;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned table1 (unsigned margin, char const * string, struct column * column) 

{ 
	extern unsigned line; 
	extern char const * style; 
	extern char const * style_empty; 
	extern char const * label_empty; 
	struct column * object; 
	line = 1; 
	headers (column); 
	indent (margin++, "<table class='%s'>", style); 
	indent (margin++, "<tr class='%s'>", style); 
	indent (margin++, "<th class='%s'>", style); 
	indent (margin, "%s", column->name); 
	indent (margin--, "</th>"); 
	for (object = column -> next; object; object = object -> next) 
	{ 
		indent (margin++, "<th class='%s'>", style); 
		indent (margin, "%s", object->name); 
		indent (margin--, "</th>"); 
	} 
	indent (margin--, "</tr>"); 
	for (columns (column); !feof (stdin); columns (column)) 
	{ 
		indent (margin++, "<tr class='%s'>", style); 
		indent (margin++, "<td class='%s'>", style); 
		indent (margin, "%s", column->name); 
		indent (margin--, "</td>"); 
		for (object = column -> next; object; object = object -> next) 
		{ 
			indent (margin++, "<td class='%s'>", object -> style); 
			indent (margin, "%s", object -> label); 
			indent (margin--, "</td>"); 
			object->style = style_empty; 
			object->label = label_empty; 
		} 
		indent (margin--, "</tr>"); 
	} 
	indent (margin--, "</table>"); 
	posted (margin); 
	finish (column); 
	return (margin); 
} 

/*====================================================================*
 *
 *   void table2 (unsigned margin, char const * string, struct column * column);
 *
 *   print table in DOCBOOK format;
 *
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

static unsigned table2 (unsigned margin, char const * string, struct column * column) 

{ 
	extern unsigned line; 
	extern unsigned cols; 
	extern char const * style_empty; 
	extern char const * label_empty; 
	struct column * object; 
	line = 1; 
	headers (column); 
	indent (margin++, "<table>"); 
	indent (margin++, "<title>"); 
	indent (margin, "%s", string); 
	indent (margin--, "</title>"); 
	indent (margin++, "<tgroup cols='%d'>", cols); 
	indent (margin, "<colspec colname='%s'/>", column -> name); 
	for (object = column -> next; object; object = object -> next) 
	{ 
		indent (margin, "<colspec colname='%s'/>", object -> name); 
	} 
	indent (margin++, "<thead>"); 
	indent (margin++, "<row>"); 
	indent (margin++, "<entry>"); 
	indent (margin, "%s", column -> name); 
	indent (margin--, "</entry>"); 
	for (object = column -> next; object; object = object -> next) 
	{ 
		indent (margin++, "<entry>"); 
		indent (margin, "%s", object -> name); 
		indent (margin--, "</entry>"); 
	} 
	indent (margin--, "</row>"); 
	indent (margin--, "</thead>"); 
	indent (margin++, "<tbody>"); 
	for (columns (column); !feof (stdin); columns (column)) 
	{ 
		indent (margin++, "<row>"); 
		indent (margin++, "<entry>"); 
		indent (margin, "%s", column -> name); 
		indent (margin--, "</entry>"); 
		for (object = column -> next; object; object = object -> next) 
		{ 
			indent (margin++, "<entry>"); 
			indent (margin, "%s", object -> label); 
			indent (margin--, "</entry>"); 
			object->style = style_empty; 
			object->label = label_empty; 
		} 
		indent (margin--, "</row>"); 
	} 
	indent (margin--, "</tbody>"); 
	indent (margin--, "</tgroup>"); 
	indent (margin--, "</table>"); 
	finish (column); 
	return (margin); 
} 

/*====================================================================*
 *
 *   int main (int argc, char const * argv []);
 *   
 *   
 *.  Motley Tools by Charles Maier;
 *:  Copyright (c) 2001-2006 by Charles Maier Associates Limited;
 *;  Licensed under the Internet Software Consortium License;
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv []) 

{ 
	static char const * optv [] = 
	{ 
		"de:hl:m:st:", 
		PUTOPTV_S_FUNNEL, 
		"print sparse matrix table", 
		"d\tprint DOCBOOK table on stdout", 
		"e s\tempty label is (s) [" LITERAL (EMPTY) "]", 
		"h\tprint HTML table on stdout", 
		"l n\tindent level is (n) [" LITERAL (LEVEL) "]", 
		"m s\tmatch label is (s) [" LITERAL (MATCH) "]", 
		"s\tprint CSS2 stylesheet on stdout", 
		"t s\ttitle is (s) [" LITERAL (TITLE) "]", 
		(char const *)(0)
	}; 
	unsigned (* table) (unsigned, char const *, struct column *) = table1; 
	unsigned margin = LEVEL; 
	char const * header = TITLE; 
	struct column column = 
	{ 
		(struct column *) (0), 
		(char const *)(0), 
		(char const *)(0), 
		(char *)(0)
	}; 
	signed c; 
	while (~ (c = getoptv (argc, argv, optv))) 
	{ 
		switch (c) 
		{ 
		case 'd': 
			table = table2; 
			break; 
		case 'e': 
			label_empty = optarg; 
			break; 
		case 'h': 
			table = table1; 
			break; 
		case 'l': 
			margin = (unsigned)(uintspec (optarg, 0, 16)); 
			break; 
		case 'm': 
			label_match = optarg; 
			break; 
		case 's': 
			stylesheet (margin); 
			return (0); 
		case 't': 
			header = optarg; 
			break; 
		default: 
			break; 
		} 
	} 
	argc -= optind; 
	argv += optind; 
	if (!argc) 
	{ 
		margin = table (margin, header, & column); 
	} 
	while ((argc) && (* argv)) 
	{ 
		if (efreopen (* argv, "rb", stdin)) 
		{ 
			margin = table (margin, header, & column); 
		} 
		argc--; 
		argv++; 
	} 
	return (0); 
} 

